crates/re_sdk/src/global.rs

```diff
@@ -0,0 +1,360 @@
+//! Keeps track of global and thread-local [`RecordingStream`]s and handles fallback logic between
+//! them.
+
+use std::cell::RefCell;
+
+use once_cell::sync::OnceCell;
+use parking_lot::RwLock;
+
+use crate::RecordingStream;
+
+// ---
+
+// TODO: use Jeremy's
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+#[allow(missing_docs)]
+pub enum RecordingType {
+    Unknown,
+    Data,
+    Blueprint,
+}
+
+impl std::fmt::Display for RecordingType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(match self {
+            RecordingType::Unknown => "unknown",
+            RecordingType::Data => "data",
+            RecordingType::Blueprint => "blueprint",
+        })
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+enum RecordingScope {
+    Global,
+    ThreadLocal,
+}
+
+impl std::fmt::Display for RecordingScope {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(match self {
+            RecordingScope::Global => "global",
+            RecordingScope::ThreadLocal => "thread-local",
+        })
+    }
+}
+
+// ---
+
+static GLOBAL_DATA_RECORDING: OnceCell<RwLock<Option<RecordingStream>>> = OnceCell::new();
+thread_local! {
+    static LOCAL_DATA_RECORDING: RefCell<Option<RecordingStream>> = RefCell::new(None);
+}
+
+static GLOBAL_BLUEPRINT_RECORDING: OnceCell<RwLock<Option<RecordingStream>>> = OnceCell::new();
+thread_local! {
+    static LOCAL_BLUEPRINT_RECORDING: RefCell<Option<RecordingStream>> = RefCell::new(None);
+}
+
+impl RecordingStream {
+    /// Returns `overrides` if it exists, otherwise returns the most appropriate active recording
+    /// of the specified type (i.e. thread-local first, then global scope), if any.
+    #[inline]
+    pub fn get(
+        which: RecordingType,
+        overrides: Option<RecordingStream>,
+    ) -> Option<RecordingStream> {
```

Comment:
This seems like weird pattern... is this just to replace:
```
some_recording.or_else(RecordingStream::get(RecordingType::Default))
```
with:
```
RecordingStream::get(RecordingType::Default, some_recording);
```

Reply:
Yes, we do this all over the place